sdks/python/apache_beam/runners/portability/portable_stager_test.py

```diff
@@ -0,0 +1,148 @@
+
+  def tearDown(self):
+    if self._temp_dir:
+      shutil.rmtree(self._temp_dir)
+    if self._remote_dir:
+      shutil.rmtree(self._remote_dir)
+
+  def stage_files(self, files):
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+    beam_artifact_api_pb2_grpc.add_ArtifactStagingServiceServicer_to_server(
+        TestLocalFileSystemArtifactStagingServiceServicer(self._remote_dir),
+        server)
+    test_port = server.add_insecure_port('[::]:0')
+    server.start()
+    stager = portable_stager.PortableStager(
+        grpc.insecure_channel('localhost:%s' % test_port))
+    for from_file, to_file in files:
+      stager.stage_artifact(
+          local_path_to_artifact=os.path.join(self._temp_dir, from_file),
+          artifact_name=to_file)
+
+    return stager._artifacts
+
+  def test_stage_single_file(self):
+    from_file = 'test_local.txt'
+    to_file = 'test_remote.txt'
+
+    with open(os.path.join(self._temp_dir, from_file), 'wb') as f:
+      f.write(b'abc')
+
+    copied_files = self.stage_files([('test_local.txt', 'test_remote.txt')])
+    self.assertTrue(
+        filecmp.cmp(
+            os.path.join(self._temp_dir, from_file),
+            os.path.join(self._remote_dir, to_file)))
+    self.assertEqual([to_file], [manifest.name for manifest in copied_files])
+
+  def test_stage_multiple_files(self):
+
+    files = [
+        ('test_local_100.txt', 'test_remote_100.txt', 100, 's'),  #
+        ('test_local_100.binary', 'test_remote_100.binary', 100, 'b'),  #
+        ('test_local_1k.txt', 'test_remote_1k.txt', 1 << 10, 's'),  #
+        ('test_local_1k.binary', 'test_remote_1k.binary', 1 << 10, 'b'),  #
+        ('test_local_1m.txt', 'test_remote_1m.txt', 1 << 20, 's'),
+        ('test_local_1m.binary', 'test_remote_1m.binary', 1 << 20, 'b'),
+        ('test_local_10m.txt', 'test_remote_10m.txt', 10 * (1 << 20), 's'),
+        ('test_local_10m.binary', 'test_remote_10m.binary', 10 * (1 << 20), 'b')
+    ]
+
+    for (from_file, _, size, type) in files:
+      chars = list(string.printable)
+      random.shuffle(chars)
+      chars = list(int(size / len(chars)) * chars + chars[0:size % len(chars)])
+      if type == 's':
+        with open(
+            os.path.join(self._temp_dir, from_file), 'w',
+            buffering=2 << 22) as f:
+          f.write(''.join(chars))
+      if type == 'b':
+        with open(
+            os.path.join(self._temp_dir, from_file), 'wb',
+            buffering=2 << 22) as f:
+          f.write(''.join(chars))
+
+    copied_files = self.stage_files(
+        [(from_file, to_file) for (from_file, to_file, _, _) in files])
+
+    for from_file, to_file, _, _ in files:
+      ff = os.path.join(self._temp_dir, from_file)
+      rf = os.path.join(self._remote_dir, to_file)
+      self.assertTrue(
+          filecmp.cmp(ff, rf),
+          'Local file {0} and remote file {1} are not the same.'.format(ff, rf))
+    self.assertEqual([to_file for _, to_file, _, _ in files],
```

Comment:
Same comment applies here - `manifest` is not the best name here. Also, consider sorting the list of items when comparing the manifest, if the order is not enforced.

Reply:
Good point. Added the sorting.